src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/RequireThisCheck.java

```diff
@@ -756,11 +756,11 @@ private static boolean isUserDefinedArrangementOfThis(AbstractFrame currentFrame
         final DetailAST blockFrameNameIdent = currentFrame.getFrameNameIdent();
         final DetailAST definitionToken = blockFrameNameIdent.getParent();
         final DetailAST blockStartToken = definitionToken.findFirstToken(TokenTypes.SLIST);
-        final DetailAST blockEndToken = getBlockEndToken(blockFrameNameIdent, blockStartToken);
+        final DetailAST blockEndToken = getBlockEndToken(blockStartToken);
 
         boolean userDefinedArrangementOfThis = false;
 
-        final Set<DetailAST> variableUsagesInsideBlock =
+        final List<DetailAST> variableUsagesInsideBlock =
             getAllTokensWhichAreEqualToCurrent(definitionToken, ident,
```

Comment:
Random arrangements of elements in the `Set` was causing this. There will be no performance hit as the `Set`(`HashSet`) was being traversed every time, none of the properties of hashing were used, if anything TC will be reduced (No HashTable made)

This applies to all the places where `Set` is replaced by `List`

Reply:
>Random arrangements of elements in the Set was causing this.

Unfortunately this a reason to disable pitest mutation or keep suppression forever.
We can not damage logic of Check to make some pitest happy.
All tools have mistakes and false positives, pitest is one of them. Until now we always find a way to change code to not really damage logic. But we were on age few times, and some time did updates in favor of pitest. But it should not be our primary goal to make pitest happy. our code should be logical and robust.

We need to rethink this. Or I need more details on survivals.

Reply:
>  We can not damage the logic of Check to make some pitest happy.

It is not being damaged, replacing Sets with Lists shouldn't be a problem, we were actually using Sets as lists earlier also, none of the hashing properties were being utilized, it was just elements arranged randomly.  All the usages of sets were inside for loops. If a regression appears in the future, it will be investigated and fixed.

I don't think this is a false positive and I feel pitest did a good job, RequireThis uses frames, and frames are added/removed in the order in which they are present in code, randomly arranging them just for traversing them later is never a good idea.

> We need to rethink this. Or I need more details on survivals.

Survivals after changing sets to lists?

Reply:
set does not allow us to put same ast twice, this is main criteria why this collection is selected in many places.

Reply:
How does random order in set cause an issue for killing mutations?

Reply:
> set does not allow us to put same ast twice, this is main criteria why this collection is selected in many places.

Why would there be duplicates? none of the algorithms that involve updating the set visit the same ast twice. I even checked it explicitly by modifying the code to compare elements in `list` and in `new HashSet<>(list)`, they were always the same, went through the algorithms manually to check if such case is possible, it is not.
Please share a case where you think there will be duplicate entries.

Reply:
> How does random order in set cause an issue for killing mutations?

The reason is similar to https://github.com/checkstyle/checkstyle/issues/11427#issuecomment-1102539871.

Reply:
@Vyom-Yadav could we use a `LinkedHashSet` here? Then, we avoid duplicates and also maintain order.

Reply:
I had originally planned to use that only, used it in prev PR, https://github.com/checkstyle/checkstyle/pull/11551#pullrequestreview-946568511, but it was suggested to not use it if hash benefits are not availed, I don't see how can the list end up having duplicates but I am ok with using LinkedHashSet to be on the safer side if everyone agrees.

Reply:
Why we need order ?

Reply:
> Why we need order?

The mutation can be killed without order also but shouldn't be IMO. Also, why pay extra for randomness when we can get the same result from using it in an ordered way.
The scenario-
```
final List<DetailAST> returnsInsideBlock = getAllTokensOfType(definitionToken,
    TokenTypes.LITERAL_RETURN, blockEndToken.getLineNo());                    

for (DetailAST returnToken : returnsInsideBlock) {      
    returnedVariable = isAstInside(returnToken, ident); 
    if (returnedVariable) { // Mutation is here, replaced it with false, never breaking from the loop                             
        break;                                          
    }                                                   
}                                                       
```
Now if we never break, there are cases when the latter `returnToken` won't be inside the tree and `returnedVariable` will return false, which will override the previous value of it being `true`, this can be killed if we add a condition to only change `returnedVariable` when it false. 

This way we will traverse the whole loop irrespective of whether we have found our result or not.
This isn't very good in my opinion. Another reason for replacing sets, which I mentioned earlier too is that we don't avail any hash benefits, duplication is not possible IMO, please share if you think duplication is possible.

Reply:
This can be written with Stream API:

```java
        return returnsInsideBlock.stream()
                .anyMatch(returnToken -> isAstInside(returnToken, ident));
```

Reply:
@Vyom-Yadav ping

Reply:
> This can be written with Stream API:

Yes, that will work, we can discuss replacing sets with lists in some other issue.